File: src/main/scala/example/expression/j/e1.scala
package example.expression.j  /*DD:LD:AI*/

import com.github.javaparser.ast.body.MethodDeclaration
import example.expression.domain.{Evolution, M1, MathDomain}
import org.combinators.templating.twirl.Java

/**
  * Truly independent of the specific design solution.
  *
  * Still Java-based, naturally and JUnit
  */
trait e1 extends Evolution with JavaGenerator with JUnitTestGenerator with M1 {
  self:e0 =>
  val domain:MathDomain

  abstract override def logic(exp:domain.Atomic)(op:domain.Operation): Seq[Statement] = {
    val subs = subExpressions(exp)
    // generate the actual body
    op match {
      case Eval =>
        exp match {
          case Sub => Java(s"return ${dispatch(subs(domain.base.left), Eval)} - ${dispatch(subs(domain.base.right), Eval)};").statements()
          case _ => super.logic(exp)(op)
        }

      case _ => super.logic(exp)(op)
    }
  }

  abstract override def testGenerator: Seq[MethodDeclaration] = {
    super.testGenerator :+ testMethod(M1_tests)
  }
}

File: attic/oldjava/e1.java
package org.combinators.ep.language.java;

class e1 {
// GOING AWAY
}
///**
//  * Truly independent of the specific design solution.
//  *
//  * Still Java-based, naturally and JUnit
//  */
//trait e1 extends Evolution with DomainIndependentJavaGenerator with JUnitTestGenerator with M1 {
//  self:e0 =>
//  import domain._
//  abstract override def logic(exp:DataType, op:Operation): Seq[Statement] = {
//    op match {
//      case Eval =>
//        exp match {
//          case Sub => result(Java(s"${dispatch(expression(exp, base.left), op)} - ${dispatch(expression(exp, base.right), op)}").expression())
//          case _ => super.logic(exp, op)
//        }
//
//      case _ => super.logic(exp, op)
//    }
//  }
//
//  abstract override def testGenerator: Seq[MethodDeclaration] = {
//    super.testGenerator ++ testMethod(M1_tests)
//  }
//}
